### PR TITLE
[UPG-NAT] Controlled CG-NAT

### PR DESCRIPTION
**Pull request by sergeymatov**
In order to allow UPG perform CG-NAT function following changes has been introduced:
To UPG:

- Ability to allocate and manage NAT Pools
- UE IP Address pool information IE to communicate NAT Pools to SMC
- BBF PFCP IEs to manage NAT Pool lookup and applying NAT to given UE IP
- UPG can manage NAT bindings configured for VPP NAT plugin

To NAT plugin:

- NAT bindings allocation/deletion mechanics exposed to UPG
- Binding lookup via bihash
- Override ED-NAT slow-path address and port allocation in respect to NAT binding created for given user address

### CODE REVIEW

File: upf/pfcp.c [hunk 1]
```diff
@@ -4674,6 +4703,135 @@ encode_alternative_smf_ip_address (void *p, u8 ** vec)
...
+}
+
+//BBF UE Function Features
+static u8 *
+format_bbf_up_function_features (u8 * s, va_list * args)
+{
+  pfcp_bbf_up_function_features_t *n =
+    va_arg (*args, pfcp_bbf_up_function_features_t *);
+
+  s = format (s, "NAT:%u", (*n & BBF_UP_NAT));
```
Comment (RoadRunnr) [inline upf/pfcp.c:4796]:
```suggestion
  s = format (s, "NAT:%u", !!(*n & BBF_UP_NAT));
```

File: upf/pfcp.c [hunk 2]
```diff
@@ -4674,6 +4703,135 @@ encode_alternative_smf_ip_address (void *p, u8 ** vec)
...
+
+}
+
+static int
+encode_bbf_up_function_features (void *p, u8 ** vec)
+{
+  u32 *v = p;
+
+  put_u16_little (*vec, *v);
+  put_u16_little (*vec, *v >> 16);
```
Comment (RoadRunnr) [inline upf/pfcp.c:4825]:
```suggestion
  put_u32_little (*vec, *v);
```

File: upf/pfcp.c [hunk 3]
```diff
@@ -6003,6 +6171,48 @@ static struct pfcp_group_ie_def pfcp_update_access_forwarding_action_information
...
+
+static u8 *
+format_ip_version (u8 * s, va_list * args)
+{
+  pfcp_ip_version_t *v = va_arg (*args, pfcp_ip_version_t *);
+
+  if (ISSET_BIT(*v, IP_VERSION_4))
+    s = format (s, "IPv4");
+  else
+    s = format (s, "IPv6");
```
Comment (RoadRunnr) [inline upf/pfcp.c:6211]: both flags can be set (v4 and v6)

File: upf/pfcp.c [hunk 4]
```diff
@@ -6003,6 +6171,48 @@ static struct pfcp_group_ie_def pfcp_update_access_forwarding_action_information
...
+      .type = PFCP_IE_BBF_NAT_PORT_BLOCK,
+      .vendor = VENDOR_BBF,
+      .is_array = true,
+      .offset = offsetof(pfcp_ue_ip_address_pool_information_t,
+                         port_blocks)
+    },
+  };
+
+#define encode_ip_version encode_u8_ie
+#define decode_ip_version decode_u8_ie
```
Comment (RoadRunnr) [inline upf/pfcp.c:6201]: similar IEs implement some minimal error checking on the flags, e.g. check that at least one bit is set.

File: upf/pfcp.h [hunk 5]
```diff
@@ -2329,7 +2384,7 @@ typedef struct
     pfcp_user_plane_ip_resource_information_t
     * user_plane_ip_resource_information;
   pfcp_tp_build_id_t tp_build_id;
-  pfcp_ue_ip_address_pool_identity_t *ue_ip_address_pool_identity;
+  pfcp_ue_ip_address_pool_identity_t ue_ip_address_pool_identity;
```
Comment (RoadRunnr) [inline upf/pfcp.h:2387]: * this is wrong, should be `UE IP address Pool Information`
* please keep the fields in the order from the specification and 3GPP IEs before the vendor IEs

File: upf/pfcp.h [hunk 6]
```diff
@@ -2360,7 +2415,7 @@ typedef struct
     pfcp_user_plane_ip_resource_information_t
     * user_plane_ip_resource_information;
   pfcp_pfcpaureq_flags_t pfcpaureq_flags;
-  pfcp_ue_ip_address_pool_identity_t *ue_ip_address_pool_identity;
+  pfcp_ue_ip_address_pool_identity_t ue_ip_address_pool_identity;
```
Comment (RoadRunnr) [inline upf/pfcp.h:2418]: * this is wrong, should be `UE IP address Pool Information`
* please keep the fields in the order from the specification and 3GPP IEs before the vendor IEs

File: upf/pfcp.h [hunk 7]
```diff
@@ -2402,10 +2479,12 @@ typedef struct
   pfcp_recovery_time_stamp_t recovery_time_stamp;
   pfcp_cp_function_features_t cp_function_features;
   pfcp_up_function_features_t up_function_features;
+  pfcp_bbf_up_function_features_t bbf_up_function_features;
```
Comment (RoadRunnr) [inline upf/pfcp.h:2482]: vendor IEs should go after 3GPP IEs

File: upf/upf.c [hunk 8]
```diff
@@ -50,6 +51,134 @@
...
+    return NULL;
+
+  np = pool_elt_at_index (gtm->nat_pools, p[0]);
+
+  return np;
+}
+
+int
+upf_init_nat_addresses (upf_nat_pool_t * np, ip4_address_t * start_addr,
+			ip4_address_t * end_addr)
```
Comment (RoadRunnr) [inline upf/upf.c:108]: why are the addresses passed by reference? the IP is small enough to be passed by value.

File: upf/upf.c [hunk 9]
```diff
@@ -50,6 +51,134 @@
...
+  do
+    {
+      upf_nat_addr_t *ap;
+      vec_add2 (np->addresses, ap, 1);
+      ap->ext_addr = addr;
+      ap->used_blocks = 0;
+      increment_v4_address (&addr);
+      ++i;
+    }
+  while (i < count);
```
Comment (RoadRunnr) [inline upf/upf.c:130]: something like this might be simpler:
```suggestion
  for (i = clib_net_to_host_u32 (addr.as_u32); 
       i < clib_net_to_host_u32 (end.as_u32); i++)
    {
      upf_nat_addr_t *ap;
      vec_add2 (np->addresses, ap, 1);
      ap->ext_addr.as_u32 = clib_host_to_net_u32(i)
      ap->used_blocks = 0;
    };
```

File: upf/upf.c [hunk 10]
```diff
@@ -50,6 +51,134 @@
...
+  return 0;
+}
+
+int
+vnet_upf_nat_pool_add_del (u8 * nwi_name, ip4_address_t * start_addr,
+			   ip4_address_t * end_addr, u8 * name,
```
Comment (RoadRunnr) [inline upf/upf.c:136]: the IPs can be passed by value

File: upf/upf_pfcp.c [hunk 11]
```diff
@@ -1020,6 +1021,17 @@ pfcp_disable_session (upf_session_t * sx)
   return 0;
 }
 
+void
+upf_delete_nat_binding (upf_session_t * sx)
+{
+  upf_nat_addr_t *ap = sx->nat_addr;
+  if (ap->used_blocks > 0)
```
Comment (RoadRunnr) [inline upf/upf_pfcp.c:1028]: calling this with `ap->used_blocks == 0` would be critical error, right?
Then maybe this should be an assert,

File: upf/upf_pfcp.c [hunk 12]
```diff
@@ -1020,6 +1021,17 @@ pfcp_disable_session (upf_session_t * sx)
...
+    ap->used_blocks -= 1;
+  upf_nat_del_binding =
```
Comment (RoadRunnr) [inline upf/upf_pfcp.c:1030]: wouldn't it make more sense to resolve the symbol only once during initialization?

File: upf/upf_pfcp_api.c [hunk 13]
```diff
@@ -147,6 +149,41 @@ init_tp_error_report (pfcp_tp_error_report_t * report,
...
+  upf_main_t *gtm = &upf_main;
+  upf_nat_pool_t *np;
+  upf_ueip_pool_info_t *ue_p;
+
+  vec_alloc (*ue_pool_info, pool_elts (gtm->ueip_pools));
+
+  pool_foreach (ue_p, gtm->ueip_pools)
+  {
+    pfcp_ue_ip_address_pool_information_t *ueif;
+    vec_add2 (*ue_pool_info, ueif, 1);
```
Comment (RoadRunnr) [inline upf/upf_pfcp_api.c:165]: there should be empty line after the local variable definition

File: upf/upf_pfcp_api.c [hunk 14]
```diff
@@ -147,6 +149,41 @@ init_tp_error_report (pfcp_tp_error_report_t * report,
...
+
+    ueif->network_instance = vec_dup (ue_p->nwi_name);
+    SET_BIT (ueif->grp.fields,
+	     UE_IP_ADDRESS_POOL_INFORMATION_NETWORK_INSTANCE);
+
+    pool_foreach (np, gtm->nat_pools)
+    {
+      if (!(vec_is_equal (np->network_instance, ue_p->nwi_name)))
+	continue;
+      pfcp_bbf_nat_port_block_t *block;
```
Comment (RoadRunnr) [inline upf/upf_pfcp_api.c:177]: local variable definition should not follow code

File: upf/upf_pfcp_api.c [hunk 15]
```diff
@@ -687,6 +731,66 @@ validate_ue_ip (upf_session_t * sx, upf_nwi_t * nwi,
     response->failed_rule_id.id = pdr->pdr_id;				\
   } while (0)
 
+upf_nat_addr_t *
+get_nat_addr (upf_nat_pool_t * np)
+{
+  upf_nat_addr_t *this_addr = NULL, *addr = NULL;
+  u32 least_locked_ports = ~0;
+  vec_foreach (this_addr, np->addresses)
```
Comment (RoadRunnr) [inline upf/upf_pfcp_api.c:739]: this could be very expensive when using large pools

File: upf/upf_pfcp_api.c [hunk 16]
```diff
@@ -687,6 +731,66 @@ validate_ue_ip (upf_session_t * sx, upf_nwi_t * nwi,
...
+	  SET_BIT (created_binding->grp.fields,
+		   TP_CREATED_BINDING_NAT_OUTSIDE_ADDRESS);
+	  SET_BIT (created_binding->grp.fields,
+		   TP_CREATED_BINDING_NAT_EXTERNAL_PORT_RANGE);
+	  return 0;
+	}
+      port_start += np->port_block_size;
+      port_end = port_start + np->port_block_size;
+    }
+  while (port_end < np->max_port);
```
Comment (RoadRunnr) [inline upf/upf_pfcp_api.c:789]: I hate this trial and error approach to finding a suitable block

**Review by RoadRunnr — CHANGES_REQUESTED** [review]

File: upf/upf_pfcp_api.c (hunk 16 above)
Comment (sergeymatov) [reply upf/upf_pfcp_api.c]: I've removed this try-err thing out from UPG code and make logic for finding port block inside NAT code.
NAT plugin in full of such constructions in terms of port selection. Now its a part of binding validation procedure.

File: upf/upf_pfcp_api.c (hunk 15 above)
Comment (sergeymatov) [reply upf/upf_pfcp_api.c:742]: Yes, it's true. For phase 2 we have I assumed to have simple pool management algorithm (but still it handles more than 1K establishment requests per sec on a /18 networks)

File: upf/upf_pfcp_api.c (hunk 14 above)
Comment (sergeymatov) [reply upf/upf_pfcp_api.c:178]: done

File: upf/upf_pfcp_api.c (hunk 13 above)
Comment (sergeymatov) [reply upf/upf_pfcp_api.c:166]: done

File: upf/upf_pfcp.c (hunk 12 above)
Comment (sergeymatov) [reply upf/upf_pfcp.c:1033]: Yes, it would. But it turned out that pointer to a function loosing reference.
In VPP code is general thing to resolve symbol right before calling the function.

File: upf/pfcp.c (hunk 4 above)
Comment (sergeymatov) [reply upf/pfcp.c]: introduced enc/dec functions for this, thanks

File: upf/pfcp.h (hunk 5 above)
Comment (sergeymatov) [reply upf/pfcp.h]: fixed

File: upf/pfcp.h (hunk 6 above)
Comment (sergeymatov) [reply upf/pfcp.h]: fixed

File: upf/pfcp.h [hunk 17]
```diff
@@ -2377,6 +2432,26 @@ typedef struct
...
+
+typedef struct
+{
+  struct pfcp_group grp;
+
+  pfcp_ue_ip_address_pool_identity_t ue_ip_address_pool_identity;
+  pfcp_network_instance_t network_instance;
+  pfcp_ip_version_t ip_version;
+  pfcp_bbf_nat_port_block_t *port_blocks;
+} pfcp_ue_ip_address_pool_information_t;
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/pfcp.h]: thanks, moved declaration

File: upf/pfcp.h [hunk 18]
```diff
@@ -2385,11 +2460,13 @@ enum
   ASSOCIATION_PROCEDURE_RESPONSE_RECOVERY_TIME_STAMP,
   ASSOCIATION_PROCEDURE_RESPONSE_UP_FUNCTION_FEATURES,
   ASSOCIATION_PROCEDURE_RESPONSE_CP_FUNCTION_FEATURES,
+  ASSOCIATION_PROCEDURE_RESPONSE_BBF_UP_FUNCTION_FEATURES,
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/pfcp.h:2463]: moved

File: upf/pfcp.h [hunk 19]
```diff
@@ -2385,11 +2460,13 @@ enum
...
   ASSOCIATION_PROCEDURE_RESPONSE_USER_PLANE_IP_RESOURCE_INFORMATION,
   ASSOCIATION_PROCEDURE_RESPONSE_TP_BUILD_ID,
   ASSOCIATION_PROCEDURE_RESPONSE_UE_IP_ADDRESS_POOL_IDENTITY,
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/pfcp.h]: done

File: upf/pfcp.h [hunk 20]
```diff
@@ -2402,10 +2479,12 @@ typedef struct
   pfcp_recovery_time_stamp_t recovery_time_stamp;
   pfcp_cp_function_features_t cp_function_features;
   pfcp_up_function_features_t up_function_features;
+  pfcp_bbf_up_function_features_t bbf_up_function_features;
     pfcp_user_plane_ip_resource_information_t
     * user_plane_ip_resource_information;
   pfcp_tp_build_id_t tp_build_id;
-  pfcp_ue_ip_address_pool_identity_t *ue_ip_address_pool_identity;
+  pfcp_ue_ip_address_pool_identity_t ue_ip_address_pool_identity;
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/pfcp.h]: done

File: upf/pfcp.h (hunk 7 above)
Comment (sergeymatov) [reply upf/pfcp.h]: moved

File: upf/upf.c (hunk 8 above)
Comment (sergeymatov) [reply upf/upf.c]: fixed

File: upf/upf.c (hunk 9 above)
Comment (sergeymatov) [reply upf/upf.c]: thanks, fixed
RE bytes ordering - it seems NAT plugin code has an issue, cause this approach is inspired by NAT Cli

File: upf/upf.c (hunk 10 above)
Comment (sergeymatov) [reply upf/upf.c]: done

File: upf/upf.h [hunk 21]
```diff
@@ -340,6 +340,30 @@ typedef struct
...
+  u32 vrf_id;
+  u16 max_blocks_per_addr;
+  u8 *network_instance;
+} upf_nat_pool_t;
+
+typedef struct
+{
+  u8 *identity;
+  u8 *nwi_name;
+} upf_ueip_pool_info_t;
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/upf.h]: renamed, thanks

File: upf/upf_pfcp.c (hunk 11 above)
Comment (sergeymatov) [reply upf/upf_pfcp.c]: added

File: upf/pfcp.c (hunk 3 above)
Comment (sergeymatov) [reply upf/pfcp.c]: fixed

File: upf/pfcp.c (hunk 2 above)
Comment (sergeymatov) [reply upf/pfcp.c]: oh, missed this. thanks

File: upf/pfcp.c [hunk 22]
```diff
@@ -4674,6 +4703,135 @@ encode_alternative_smf_ip_address (void *p, u8 ** vec)
...
+  s = format (s, "NAT:%u", (*n & BBF_UP_NAT));
+  return s;
+}
+
+static int
+decode_bbf_up_function_features (u8 * data, u16 length, void *p)
+{
+  u64 *v = p;
+
+  if (length % 2 != 0 || length < 2)
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/pfcp.c]: check line updated

File: upf/pfcp.c [hunk 23]
```diff
@@ -4674,6 +4703,135 @@ encode_alternative_smf_ip_address (void *p, u8 ** vec)
...
+
+  if (length < 4)
+    return PFCP_CAUSE_INVALID_LENGTH;
+
+  v->start_port = get_u16 (data);
+  length -= 2;
+
+  if (length < 2)
+    return PFCP_CAUSE_INVALID_LENGTH;
+
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/pfcp.c]: removed

File: upf/test/test_upf.py [hunk 24]
```diff
@@ -359,7 +367,6 @@ def associate(self):
             IE_NodeId(id_type="FQDN", id="ergw")
         ]), PFCPAssociationSetupResponse)
         self.assertEqual(CauseValues[resp[IE_Cause].cause], "Request accepted")
-        self.assertIn(b"vpp", resp[IE_EnterpriseSpecific].data)
```
Comment (ivan4th) [inline upf/test/test_upf.py:362]: Is this change somehow related to this PR? This is actually a problem (#104), as the build id should contain `upg` not `vpp`, but I thought of fixing it separately and not by removing this check, but rather by fixing the id itself and replacing `vpp` with `upg` here

**Review by ivan4th — CHANGES_REQUESTED** [review]
The major problem I see is the lack of any tests (integration or e2e), but sadly this will have to wait for now.
Other than that, one of the problems I found looks a potential bug, others being some nitpicking

File: upf/test/test_upf.py (hunk 24 above)
Comment (sergeymatov) [reply upf/test/test_upf.py:362]: In general, association setup response checks should be enhanced with:
- proper build ID check
- BBF UserPlane Features checks
I think it's worth a separate review

File: upf/upf_pfcp.c [hunk 25]
```diff
@@ -1033,6 +1049,11 @@ pfcp_free_session (upf_session_t * sx)
 
   sparse_vec_free (sx->teid_by_chid);
 
+  if (sx->nat_addr)
+    {
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/upf_pfcp.c]: fixed

File: upf/upf_pfcp_api.c [hunk 26]
```diff
@@ -1319,6 +1450,23 @@ handle_create_far (upf_session_t * sx, pfcp_create_far_t * create_far,
 
 	  }
 
+	if (ISSET_BIT (far->forwarding_parameters.grp.fields,
+		       FORWARDING_PARAMETERS_BBF_APPLY_ACTION))
+	  {
+	    if (far->forwarding_parameters.bbf_apply_action &=
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/upf_pfcp_api.c]: No it's not. Thanks for notifying. Fixed

**Comment by ivan4th** [comment]
LGTM

**Review by ivan4th — APPROVED** [review]

File: upf/upf.c [hunk 27]
```diff
@@ -48,8 +49,136 @@
...
+  upf_nat_pool_t *np = NULL;
+  uword *p;
+
+  p = hash_get_mem (gtm->nat_pool_index_by_name, name);
+  if (!p)
+    return NULL;
+
+  np = pool_elt_at_index (gtm->nat_pools, p[0]);
+
+  return np;
```
Comment (RoadRunnr) [inline upf/upf.c:110]:
```suggestion
  return pool_elt_at_index (gtm->nat_pools, p[0]);
```

File: upf/upf.c [hunk 28]
```diff
@@ -48,8 +49,136 @@
...
+}
+
+int
+upf_init_nat_addresses (upf_nat_pool_t * np, ip4_address_t start_addr,
+			ip4_address_t end_addr)
+{
+  u32 i = 0;
+
+  vec_reset_length (np->addresses);
```
Comment (RoadRunnr) [inline upf/upf.c:119]: to avoid unnecessary fragmentation, you could used vec_validate to make sure the vector is large enough to hold all entries before going into the loop.

File: upf/upf_cli.c [hunk 29]
```diff
@@ -203,6 +210,118 @@ upf_name_to_labels (u8 * name)
...
+
+  return error;
+}
+
+/* *INDENT-OFF* */
+VLIB_CLI_COMMAND (upf_nat_pool_add_del_command, static) =
+{
+  .path = "upf nat pool",
+  .short_help =
+  "upf nat pool nwi <nwi-name> start <ip4-addr> end <ip4-addr> min_port <min-port> max_port <max-port> block_size <port-block-size> vrf <vrf-id> name <name> [del]",
```
Comment (RoadRunnr) [inline upf/upf_cli.c:320]: A NWI currently uniquely identifies a VRF, it the extra VRF really needed or different from the NWI VRF?

File: upf/upf_pfcp_api.c (hunk 14 above)
Comment (RoadRunnr) [reply upf/upf_pfcp_api.c:178]: this has not been addressed

**Review by RoadRunnr — CHANGES_REQUESTED** [review]

File: upf/upf_cli.c (hunk 29 above)
Comment (sergeymatov) [reply upf/upf_cli.c:320]: We are operating nwi name w/o nwi lookup itself. I will add TODO in a code, but removing vrf out will create a cascade of changes to charts.

File: upf/upf.c (hunk 28 above)
Comment (sergeymatov) [reply upf/upf.c:119]: done

File: upf/upf.c (hunk 27 above)
Comment (sergeymatov) [reply upf/upf.c:110]: done

File: upf/upf_pfcp_api.c (hunk 14 above)
Comment (sergeymatov) [reply upf/upf_pfcp_api.c:179]: Sorry, missed this here but fixed for stable branch.

**Comment by ivan4th** [comment]
LGTM from my side

File: upf/upf.c [hunk 30]
```diff
@@ -48,8 +49,135 @@
...
+		    nat_pool - gtm->nat_pools);
+
+    }
+  else
+    {
+      if (!p)
+	return VNET_API_ERROR_NO_SUCH_ENTRY;
+
+      nat_pool = pool_elt_at_index (gtm->nat_pools, p[0]);
+      //TBD: nat pool cleanup upf_delete_nat_addresses(nat_pool);
```
Comment (RoadRunnr) [inline upf/upf.c:172]: removing a NAT pool that used entries would crash at some, right?

File: upf/upf_cli.c (hunk 29 above)
Comment (RoadRunnr) [reply upf/upf_cli.c:321]: "because something somewhere else would need to be redone" is not an argument for leaving this code in a inconsistent state. A TODO for this is only acceptable if it is followed up with some urgency.

**Review by RoadRunnr — CHANGES_REQUESTED** [review]
I think the CLI (nwi vs. vrf) should be sorted out before this is merged to master

File: upf/upf_cli.c [hunk 31]
```diff
@@ -287,27 +285,28 @@ upf_nat_pool_add_del_command_fn (vlib_main_t * vm,
...
-	;
-      else if (unformat (line_input, "name %_%v%_", &name))
+      // FIXME: This is a workaround to have 2 strings fetched from args
+      // There is kind of bug if two strings are present in command line
+      // and there are two appropriate unformat calls for it
+      // somehow it affects input line (wrong end-of-string indication or
+      // something like this). To avoid this, two strings could be picked
+      // in single unformat call. Probably Address Sanitaizer should be
+      // used to identify problems with unformatting input line
+      else if (unformat (line_input, "nwi %_%v%_ name %_%v%_", &nwi_s, &name))
```
Comment (RoadRunnr) [inline upf/upf_cli.c:295]: it seems that "%s" does not suffer from this problem. It would therefore be possible to scan the names as string and then convert them to vectors.
That would IMHO be a better workaround.

Comment (RoadRunnr) [reply upf/upf_cli.c:295]: Looking at the full function also makes me wonder if the fact that `name` is not initialized to zero might play a role in the problem.

Comment (sergeymatov) [reply upf/upf_cli.c]: Initiation of name vectors to 0 resolves the issue, so I've removed W/A.
Thanks for good notice

File: upf/upf_cli.c (hunk 29 above)
Comment (sergeymatov) [reply upf/upf_cli.c]: VRF removed as redundant CLI argument

**Review by RoadRunnr — APPROVED** [review]
